### PR TITLE
added bounding box prediction for links

### PR DESCRIPTION
**Pull request by alex2awesome**
Adds a new function: `hyperlinks_with_bounding_boxes.py`

End-goal of this function is to dump a list of `<a>` href links where the links (A) all point to articles and (B) have accompanying bounding boxes that are the maximal bounding box for that link. 

(A) is achieved in two ways. If, AND ONLY IF, the article is from an English site (determined by `sites.csv`), then prediction is made on the URL text using a pure-javascript implementation of a logistic regression model trained using `storysniffer`'s data. This model achieves 94.5 f1 score on a held-out test set included with `storysniffer`. If the article is NOT in English, then we basically look at the size of the URL as a heuristic to determine if it is an article or not. This method achieves an 89 f1 score on `storysniffer's` held out set**.

(B) Maximal bounding box is determined for each <a> that points to an article to be the element in the DOM tree that is the highest parent of the `<a>`-article tag that does NOT have any other `<a>`-article links in it's subtree. 

Here: [nytimes.hyperlinks-with-bb.json.zip](https://github.com/palewire/news-homepages/files/10516130/nytimes.hyperlinks-with-bb.json.zip)
 is the full output for running the function

`python newshomepages/hyperlinks_with_bounding_boxes.py nytimes`

Output dict has the following fields:

```
{
  "bounding boxes": [ # one for every article link on the page
    {
      "url": "", # a.href
      "link_text": "", # text in the <a> block (headline)
      "x": , # x position of the maximal bounding box
      "y": , # y position of the maximal bounding box
      "width": # width of the ...
      "height": # height of the ...
      "all_text": "" # all text of the maximal bounding box
    },
...
],
"page_width": # full width of the page 
"page_height":# full height of the page 
}
```

** note that `storysniffer's` heldout set is ALSO based on English articles, so this method may need to be tweaked going forward to get a more appropriate evaluation. `<a>` that do not point to articles include "furniture" links like author pages, section fronts, and other homepages.

### CODE REVIEW

**Comment by palewire** [comment]
Here's my dumb question: Why do you need the JS to be run "on the fly" in the browser?

In other circumstances, like our Drudge analysis, I have one routine that pulls all hyperlinks via a web browser and writes the results to a CSV. Then a later routine, without a browser, does the story sniffing. 

Is there a critical bug in that approach?

**Comment by alex2awesome** [comment]
So, as far as I know, there's no way to get the size/position/etc. of a DIV using Python and the HTML. It needs to be done in the browser because the browser renders the HTML with all the CSS to get the spatial layouts. 

Now, we _can_ go back and forth to some degree.... 

(1) get html (js) -> (2) get all links (python) -> (3) storysniff all links (python) -> get bounding boxes just for those links (js)

But having the whole pipeline is JS is just a bit more efficient, in my opinion.... less confusing/less back and forth/more legible.

(BTW — the Javascript and Python storysniffer models are equivalent... I've manually verified that the Logistic Regression I wrote deterministically produces the same output as the python scikit learn logistic regression. I've also written a script that takes the the sci-kit learn object and outputs a JSON of the parameters that the JS class understands, so if you want to retrain in Python at any point, you can port into JavaScript.)

**Comment by alex2awesome** [comment]
What's the concern? Is the concern that you will want to update storysniffer, in Python, and then it will become disjointed from this use-case?

**Comment by alex2awesome** [comment]
Not a dumb question.

To show another way, what this approach to maximal bounding-boxes does is the following:

bottom-up-visualization

We find every <a> link that is NOT a furniture link, and traverse as far as we can up the tree to get a DOM element that does 
NOT have any other <a> elements in it's subtree. 

The consequences of having furniture links in there at the time we calculate bounding boxes is the following: 

image (1)

So, as you can see in the image, there are summaries and other blank spaces. I can't really see a way to get suboptimal bounding boxes ad run-time and then reason about this well enough in post-processing. Because, we'd never get a bounding box around that summary text in the first place because there's not <a> in that div. 

We'd have to dive back into the HTML which, despite our best attempts, isn't preserving with 100% accuracy (see here: https://docs.google.com/spreadsheets/d/15JHXleezjTLnsz317EsvoFbVe2q8sst1cBiYabwLd_g/edit?usp=sharing for the is_proper column where we're exploring how many of these htmls we are actually preserving correctly. I still think our method is better than raw wget, but still, I don't think it's possible to 100% preserve HTML.)

So, if we don't do this at run-time, then there are sites that we won't be able to run this method on.

**Comment by alex2awesome** [comment]
Alrighty.... updated to incorporate comments on our slack. In summary:

1. merged `hyperlinks_with_bounding_boxes.py` to `hyperlinks.py`
2. get bounding boxes for ALL urls and don't do any ML story-sniffing
3. get all css-attributes for the bounded node. It's a long list

Here's a sample on nytimes:

[nytimes.hyperlinks-with-bb.json.zip](https://github.com/palewire/news-homepages/files/10561718/nytimes.hyperlinks-with-bb.json.zip)

Here's what Jacobin looks like:

jacobin

File: newshomepages/hyperlinks.py [hunk 1]
```diff
@@ -2,13 +2,127 @@
...
-from bs4 import BeautifulSoup
 from playwright.sync_api import sync_playwright
 from playwright.sync_api._generated import BrowserContext
 from retry import retry
-from rich import print
-
-from . import utils
+# from rich import print
+
+# from . import utils
```
Comment (palewire) [inline newshomepages/hyperlinks.py:10]: Is this cruft? Something else?

**Review by palewire — CHANGES_REQUESTED** [review]
Here's one tiny thing. We've also got some linting errors in the tests that should get cleaned up.

**Comment by alex2awesome** [comment]
sorry for the delay — updated!

File: newshomepages/hyperlinks.py (hunk 1 above)
Comment (alex2awesome) [reply newshomepages/hyperlinks.py]: fixed!

**Comment by palewire** [comment]
Does Playwright allow us to get any of this information natively without the need for JS? Just curious what's possible with the tool.

**Comment by alex2awesome** [comment]
It might be possible? 

Getting elements by selectors: https://playwright.dev/python/docs/api/class-locator
Getting CSS attributes by category: https://github.com/microsoft/playwright/issues/4282#issuecomment-718602677
Getting bounding boxes: https://playwright.dev/python/docs/api/class-locator#locator-bounding-box

All the pieces seem to be there for most of what we want to retrieve. I could see some bug or definitional difference (i.e. an element's position on the page, client-side or full-page? idk, just brainstorming) meaning that these methods return slightly different values.

I like the idea of JS implementations because it's more agnostic to the specific tool, but i also get that it ends up generating unweildy code that's not easily tested

**Comment by palewire** [comment]
I'm going to close this ticket due to inactivity. If you want to pick it back up, please let me know.

**Pull request closed without merging**